contracts/src/v0.8/dev/automation/upkeeps/ProxyBalanceMonitor.sol

```diff
@@ -0,0 +1,287 @@
+    if (numCanFund < needsFunding.length) {
+      assembly {
+        mstore(needsFunding, numCanFund) // resize
+      }
+    }
+    upkeepNeeded = needsFunding.length > 0;
+    performData = abi.encode(needsFunding);
+    return (upkeepNeeded, performData);
+  }
+
+  /**
+   * @notice Called by the keeper to send funds to underfunded addresses.
+   * @param performData the abi encoded list of addresses to fund
+   */
+  function performUpkeep(bytes calldata performData) external override {
+    address[] memory needsFunding = abi.decode(performData, (address[]));
+    topUp(needsFunding);
+  }
+
+  /**
+   * @notice Withdraws the contract balance in the LINK token.
+   * @param amount the amount of the LINK to withdraw
+   * @param payee the address to pay
+   */
+  function withdraw(uint256 amount, address payable payee) external onlyOwner {
+    require(payee != address(0));
+    LINK_TOKEN.transfer(payee, amount);
+    emit FundsWithdrawn(amount, payee);
+  }
+
+  /**
+   * @notice Sets the top up amount
+   */
+  function setTopUpAmount(uint256 topUpAmount) external onlyOwner returns (uint256) {
+    require(topUpAmount > 0);
+    return s_topUpAmount = topUpAmount;
+  }
+
+  /**
+   * @notice Sets the minimum balance
+   */
+  function setMinBalance(uint256 minBalance) external onlyOwner returns (uint256) {
+    require(minBalance > 0);
+    return s_minBalance = minBalance;
+  }
+
+  /**
+   * @notice Pause the contract, which prevents executing performUpkeep
+   */
+  function pause() external onlyOwner {
+    _pause();
+  }
+
+  /**
+   * @notice Unpause the contract
+   */
+  function unpause() external onlyOwner {
+    _unpause();
+  }
+
+  /**
+   * @notice Gets the list of subscription ids being watched
+   */
+  function getWatchList() external view returns (address[] memory) {
+    return s_watchList.values();
+  }
+
+  /**
+   * @notice Gets the configured top up amount
+   */
+  function getTopUpAmount() external view returns (uint256) {
+    return s_topUpAmount;
+  }
+
+  /**
+   * @notice Gets the configured minimum balance
+   */
+  function getMinBalance() external view returns (uint256) {
+    return s_minBalance;
+  }
+
+  /**
+   * @notice checks the aggregator that the provided proxy points to, and determines
+   * if it is elligible for funding
+   * @param proxyAddress the proxy to check
+   * @return bool whether the aggregator needs funding or not
+   * @return address the address of the aggregator
+   */
+  function _needsFunding(address proxyAddress) private view returns (bool, address) {
+    IAggregator aggregator;
+    IAggregatorProxy proxy = IAggregatorProxy(proxyAddress);
+    try proxy.aggregator() returns (address aggregatorAddress) {
+      aggregator = IAggregator(aggregatorAddress);
+    } catch {
+      return (false, address(0));
+    }
+    try aggregator.linkAvailableForPayment() returns (int256 balance) {
+      if (balance < 0 || uint256(balance) > s_minBalance) {
```

Comment:
This cast here: `uint256(balance)` could cause overflow if it's negative right?

Reply:
`balance < 0 ||` should cover it, not the best pattern though imo

Reply:
Oh I totally missed the first part 😄